lib/rules/function-linear-gradient-no-nonstandard-direction/index.js

```diff
@@ -37,40 +38,45 @@ const rule = function (actual) {
     }
 
     root.walkDecls(decl => {
-      functionArgumentsSearch(decl.toString().toLowerCase(), "linear-gradient", (expression, expressionIndex) => {
-        const firstArg = expression.split(",")[0].trim()
+      valueParser(decl.value).walk(valueNode => {
+        if (valueNode.type !== "function") {
+          return
+        }
+        functionArgumentsSearch(valueParser.stringify(valueNode).toString().toLowerCase(), "linear-gradient", (expression, expressionIndex) => {
+          const firstArg = expression.split(",")[0].trim()
 
-        // If the first character is a number, we can assume the user intends an angle
-        if (/[\d\.]/.test(firstArg[0])) {
-          if (/^[\d\.]+(?:deg|grad|rad|turn)$/.test(firstArg)) {
+          // If the first character is a number, we can assume the user intends an angle
+          if (/[\d\.]/.test(firstArg[0])) {
+            if (/^[\d\.]+(?:deg|grad|rad|turn)$/.test(firstArg)) {
+              return
+            }
+            complain()
             return
           }
-          complain()
-          return
-        }
 
-        // The first argument may not be a direction: it may be an angle,
-        // or a color stop (in which case user gets default direction, "to bottom")
-        // cf. https://drafts.csswg.org/css-images-3/#linear-gradient-syntax
-        if (!/left|right|top|bottom/.test(firstArg)) {
-          return
-        }
+          // The first argument may not be a direction: it may be an angle,
+          // or a color stop (in which case user gets default direction, "to bottom")
+          // cf. https://drafts.csswg.org/css-images-3/#linear-gradient-syntax
+          if (!/left|right|top|bottom/.test(firstArg)) {
+            return
+          }
 
-        const withToPrefix = !postcss.vendor.prefix(decl.value)
-        if (!isStandardDirection(firstArg, withToPrefix)) {
-          complain()
-          return
-        }
+          const withToPrefix = !postcss.vendor.prefix(valueNode.value)
+          if (!isStandardDirection(firstArg, withToPrefix)) {
+            complain()
+            return
+          }
 
-        function complain() {
-          report({
-            message: messages.rejected,
-            node: decl,
-            index: expressionIndex,
-            result,
-            ruleName,
-          })
-        }
+          function complain() {
+            report({
+              message: messages.rejected,
+              node: decl,
+              index: expressionIndex,
```

Comment:
I suspect you'll need to calculate the index based on the `expressionIndex` _and the index of the `postcss-value-parser` node_.